Source/stash.cpp

```diff
@@ -0,0 +1,839 @@
+			return -1;
+		int ii = itemId - 1;
+		rv = ii + STASH_SLOT_FIRST;
+		pi = &Stash.stashList[ii];
+	}
+
+	if (pi->isEmpty())
+		return -1;
+
+	if (pi->_itype == ItemType::Gold) {
+		int nGold = pi->_ivalue;
+		strcpy(infostr, fmt::format(ngettext("{:d} gold piece", "{:d} gold pieces", nGold), nGold).c_str());
+	} else {
+		InfoColor = pi->getTextColor();
+		if (pi->_iIdentified) {
+			strcpy(infostr, pi->_iIName);
+			PrintItemDetails(*pi);
+		} else {
+			strcpy(infostr, pi->_iName);
+			PrintItemDur(*pi);
+		}
+	}
+
+	return rv;
+}
+
+bool UseStashItem(int playerId, int cii)
+{
+	auto &player = Players[playerId];
+
+	if (player._pInvincible && player._pHitPoints == 0 && playerId == MyPlayerId)
+		return true;
+	if (pcurs != CURSOR_HAND)
+		return true;
+	if (stextflag != STORE_NONE)
+		return true;
+	if (cii < STASH_SLOT_FIRST)
+		return false;
+	if (cii > STASH_SLOT_LAST)
+		return false;
+
+	int c = cii - STASH_SLOT_FIRST;
+	Item *item = &Stash.stashList[c];
+
+	constexpr int SpeechDelay = 10;
+	if (item->IDidx == IDI_MUSHROOM) {
+		player.Say(HeroSpeech::NowThatsOneBigMushroom, SpeechDelay);
+		return true;
+	}
+	if (item->IDidx == IDI_FUNGALTM) {
+		PlaySFX(IS_IBOOK);
+		player.Say(HeroSpeech::ThatDidntDoAnything, SpeechDelay);
+		return true;
+	}
+
+	if (!AllItemsList[item->IDidx].iUsable)
+		return false;
+
+	if (!item->_iStatFlag) {
+		player.Say(HeroSpeech::ICantUseThisYet);
+		return true;
+	}
+
+	if (withdrawGoldFlag) {
+		withdrawGoldFlag = false;
+		withdrawGoldValue = 0;
+	}
+
+	if (item->IsScroll() && currlevel == 0 && !spelldata[item->_iSpell].sTownSpell) {
+		return true;
+	}
+
+	if (item->_iMiscId > IMISC_RUNEFIRST && item->_iMiscId < IMISC_RUNELAST && currlevel == 0) {
+		return true;
+	}
+
+	if (item->_iMiscId == IMISC_BOOK)
+		PlaySFX(IS_RBOOK);
+	else if (playerId == MyPlayerId)
+		PlaySFX(ItemInvSnds[ItemCAnimTbl[item->_iCurs]]);
+
+	UseItem(playerId, item->_iMiscId, item->_iSpell);
+
+	if (Stash.stashList[c]._iMiscId == IMISC_MAPOFDOOM)
+		return true;
+	if (Stash.stashList[c]._iMiscId == IMISC_NOTE) {
+		InitQTextMsg(TEXT_BOOK9);
+		IsStashOpen = false;
+		return true;
+	}
+	Stash.RemoveStashItem(c);
+
+	return true;
+}
+
+void StashStruct::RemoveStashItem(int8_t iv)
+{
+	iv++;
+
+	// Iterate through stashGrid and remove every reference to item
+	for (int8_t &itemId : Stash.stashGrids[StashPage]) {
+		if (itemId == iv || itemId == -iv) {
+			itemId = 0;
+		}
+	}
+
+	iv--;
+
+	// If the item at the end of stash array isn't the one we removed, we need to swap its position in the array with the removed item
+	if ((stashList.size() - 1) > 0 && (stashList.size() - 1) != iv) {
+		stashList[iv] = stashList[stashList.size() - 1];
+
+		for (auto &pair : Stash.stashGrids) {
+			auto &grid = pair.second;
+			for (int8_t &itemId : grid) {
+				if (itemId == stashList.size()) {
+					itemId = iv + 1;
+				}
+				if (itemId == -(stashList.size())) {
+					itemId = -(iv + 1);
+				}
+			}
+		}
+	}
+	stashList.pop_back();
+}

```

Comment:
```suggestion
	if (stashList.empty()) {
		return;
	}
	
	// If the item at the end of stash array isn't the one we removed, we need to swap its position in the array with the removed item
	std::size_t lastItemIndex = stashList.size() - 1;
	if (lastItemIndex != iv) {
		stashList[iv] = stashList[lastItemIndex];

		for (auto &pair : Stash.stashGrids) {
			auto &grid = pair.second;
			for (auto &itemId : grid) {
				if (itemId == lastItemIndex + 1) {
					itemId = iv + 1;
				}
				if (itemId == -(lastItemIndex + 1)) {
					itemId = -(iv + 1);
				}
			}
		}
	}
	stashList.pop_back();
}
```

Calling pop_back on an empty vector is undefined behaviour. Also might as well capture the intent of the vector::size() call in a named variable.